notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/TLS1_2SocketFactory.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+package com.microsoft.windowsazure.messaging.notificationhubs;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.security.KeyManagementException;
+import java.security.NoSuchAlgorithmException;
+
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
+
+import static javax.net.ssl.HttpsURLConnection.getDefaultSSLSocketFactory;
+
+/**
+ * This class forces TLS 1.2 protocol via adapter pattern.
+ */
+class TLS1_2SocketFactory extends SSLSocketFactory {
+
+    /**
+     * TLS 1.2 protocol name.
+     */
+    private static final String TLS1_2_PROTOCOL = "TLSv1.2";
+
+    /**
+     * Protocols that we allow.
+     */
+    private static final String[] ENABLED_PROTOCOLS = { TLS1_2_PROTOCOL };
+
+    /**
+     * Socket factory.
+     *
+     * Do not rename it! See https://github.com/square/okhttp/issues/2323
+     */
+    private final SSLSocketFactory delegate;
+
+    TLS1_2SocketFactory() {
+        SSLSocketFactory socketFactory = null;
+        try {
+
+            /*
+             * Explicitly specify protocol for SSL context.
+             * See https://www.java.com/en/configure_crypto.html#enableTLSv1_2
+             */
+            SSLContext sc = SSLContext.getInstance(TLS1_2_PROTOCOL);
+            sc.init(null, null, null);
+            socketFactory = sc.getSocketFactory();
+        } catch (KeyManagementException | NoSuchAlgorithmException ignored) {
+        }
+        delegate = socketFactory != null ? socketFactory : getDefaultSSLSocketFactory();
+    }
+
+    /**
+     * Force TLS 1.2 protocol on a socket.
+     *
+     * @param socket socket.
+     * @return that same socket for chaining calls.
+     */
+    private SSLSocket forceTLS1_2(Socket socket) {
+        SSLSocket sslSocket = (SSLSocket) socket;
+        sslSocket.setEnabledProtocols(ENABLED_PROTOCOLS);
```

Comment:
Alright, I've done some initial research here and it doesn't seem like Volley is a clear win. It may still be better, but more analysis should really be done than can fit into preview2. I filed a user story to take a closer look.